modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java

```diff
@@ -209,6 +223,48 @@ private JSONArray _getAssetTagsJSONArray(
 			).toArray());
 	}
 
+	private JSONObject _getAssetVocabulariesJSONObject(
+		ContentDashboardItem contentDashboardItem, Locale locale) {
+
+		List<AssetCategory> assetCategories =
+			contentDashboardItem.getAssetCategories();
+
+		Stream<AssetCategory> stream = assetCategories.stream();
+
+		Map<Long, Map<String, Object>> assetVocabularies = new HashMap<>();
+
+		stream.forEach(
+			assetCategory -> {
+				long vocabularyId = assetCategory.getVocabularyId();
+
+				Optional<Map<String, Object>> vocabularyOptional =
+					Optional.ofNullable(assetVocabularies.get(vocabularyId));
+
+				if (vocabularyOptional.isPresent()) {
```

Comment:
The idea of use Optional in this case is so we can have objects that are created just one time, so they can be inmutables.

You could implement it as something like:

Map<Long, Map<String, Object>> assetVocabularies = stream.map ...

Reply:
For example something like

Map<Long, Map<String, List<String>> assetVocabularies = new HashMap<>();

		stream.forEach(
			assetCategory -> {
				assetVocabularies.computeIfAbsent(
					vocabularyId -> {
						AssertVocabulary assetVocabulary =
							_assetVocabularyLocalService.fetchAssetVocabulary(
								vocabularyId);

						return_HashMapBuilder.<String, Object>put(
							"categories", new ArrayList<>()
						).put(
							"groupName",
							ContentDashboardGroupUtil.getGroupName(
								_groupLocalService.fetchGroup(
									assetVocabulary.getGroupId()),
								locale)
						).put(
							"isPublic",
							assetVocabulary.getVisibilityType() ==
							AssetVocabularyConstants.VISIBILITY_TYPE_PUBLIC
						).put(
							"vocabularyName",
							assetVocabulary.getTitleCurrentValue()
						).build();
					}
				);


			}).get();

Reply:
A good example of how we can use compute if absent can be found in https://www.baeldung.com/java-map-computeifabsent

Reply:
Ok, thxs, gonna try.
The `assetVocabularies` HashMap cannot be declared and assigned to the stream, right?